### PR TITLE
Azure Policy Authoring Guide

### PR DESCRIPTION
**Pull request by SenthuranSivananthan**
## Overview/Summary

New documentation to provide step-by-step guide on setting up Azure Policies using the automation.  The documentation includes:

* Built-in policy sets
  * New built-in policy set assignment
  * Remove built-in policy set assignment
* Custom policies
  * New custom policy definition
  * New custom policy set definition & assignment
  * Update custom policy definition
  * Update custom policy set definition & assignment
  * Remove custom policy definition
  * Remove custom policy set definition and assignment
  * Auto generating custom Diagnostic Settings policies for PaaS services

## This PR fixes/adds/changes/removes

1. #47 

### Breaking Changes

None

## Testing Evidence

N/A

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/azure/CanadaPubSecALZ/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/azure/CanadaPubSecALZ/issues)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/azure/CanadaPubSecALZ/tree/main)
- [x] Updated relevant and associated documentation.

### CODE REVIEW

File: docs/policy/authoring-guide.md [hunk 1]
```diff
@@ -0,0 +1,970 @@
...
+        properties: {
+          displayName: assignmentName
+          policyDefinitionId: policyScopedId
+          scope: scope
+          notScopes: [
+          ]
+          parameters: {
+            // Add any parameters identified earlier into this section
+          }
+          enforcementMode: 'Default'
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:172]: Do we want to add comment describing allowable values for enforcement mode?

File: docs/policy/authoring-guide.md [hunk 2]
```diff
@@ -0,0 +1,970 @@
...
+            "listOfMembersToIncludeInWindowsVMAdministratorsGroup": {
+                "value": "__tbd__implementation_specific__"
+            }
+        }
+    }
+    ```
+
+#### **Step 3: Update Azure DevOps Pipeline**
+
+  * Edit `.pipelines/policy.yml`
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:305]:
```suggestion
  * Edit `/.pipelines/policy.yml`
```

File: docs/policy/authoring-guide.md [hunk 3]
```diff
@@ -0,0 +1,970 @@
...
+  * Navigate to the `BuiltInPolicyJob` Job definition
+  * Navigate to the `Assign Policy Set` Step definition
+  * Add the policy assignment file name (without extension) to the `deployTemplates` array parameter
+
+#### **Step 4: Deploy built-in policy set assignment**
+
+Execute `Azure DevOps Policy pipeline` to deploy.  The policy set assignment will be deployed to the `top level management group` (i.e. `pubsec`).
+
+> It takes around 30 minutes for the assignment to be applied to the defined scope. Once it's applied, the evaluation cycle begins for resources within that scope against the newly assigned policy or initiative and depending on the effects used by the policy or initiative, resources are marked as compliant, non-compliant, or exempt. A large policy or initiative evaluated against a large scope of resources can take time. As such, there's no pre-defined expectation of when the evaluation cycle completes. Once it completes, updated compliance results are available in the portal and SDKs.  See [Azure Docs for more information](https://docs.microsoft.com/azure/governance/policy/how-to/get-compliance-data).
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:314]:
```suggestion
> It takes approximately 30 minutes for the assignment to be applied to the defined scope. Once it's applied, the evaluation cycle begins for resources within that scope against the newly assigned policy or initiative, and depending on the effects defined by the policy or initiative, resources are marked as compliant, non-compliant, or exempt. A large policy or initiative evaluated against a large scope of resources can take time. As such, there's no pre-defined expectation of when the evaluation cycle completes. Once evaluation completes, the updated compliance results are available in the portal and to SDKs.  See [Azure Docs for more information](https://docs.microsoft.com/azure/governance/policy/how-to/get-compliance-data).
```

File: docs/policy/authoring-guide.md [hunk 4]
```diff
@@ -0,0 +1,970 @@
...
+
+#### **Step 5: Verify policy set assignment**
+
+  * You can navigate to [Azure Policy Compliance][portalAzurePolicyCompliance] to verify in Azure Portal.
+  * When there are deployment errors:
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:319]:
```suggestion
  * If there are deployment errors:
```

File: docs/policy/authoring-guide.md [hunk 5]
```diff
@@ -0,0 +1,970 @@
...
+### Remove built-in policy set assignment
+
+**Steps**
+
+* [Step 1: Remove built-in policy set assignment from Azure DevOps Pipeline](#step-1-remove-built-in-policy-set-assignment-from-azure-devops-pipeline)
+* [Step 2: Remove built-in policy set assignment's IAM assignments](#step-2-remove-built-in-policy-set-assignments-iam-assignments)
+
+#### **Step 1: Remove built-in policy set assignment from Azure DevOps Pipeline**
+
+* Edit `.pipelines/policy.yml`
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:337]:
```suggestion
* Edit `/.pipelines/policy.yml`
```

File: docs/policy/authoring-guide.md [hunk 6]
```diff
@@ -0,0 +1,970 @@
...
+        }
+      }
+    }
+    ```
+
+5. Edit `azurepolicy.rules.json`
+
+    Describes the policy rule that will be evaluated by Azure Policy.  The rule can have any effect such as Audit, Deny, DeployIfNotExists.
+
+    See [Azure Policy docs for more information on creating custom policies](https://docs.microsoft.com/en-us/azure/governance/policy/concepts/definition-structure).
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:426]:
```suggestion
    See [Azure Policy docs for more information on creating custom policies](https://docs.microsoft.com/azure/governance/policy/concepts/definition-structure).
```

File: docs/policy/authoring-guide.md [hunk 7]
```diff
@@ -0,0 +1,970 @@
...
+                  value: '[parameters(\'listOfResourceTypesToAuditDiagnosticSettings\')]'
+                }
+              }
+            }
+          ]
+        }
+      }
+    ```
+
+3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This parameters JSON file is used by Azure Resource Manager (ARM) for runtime inputs.
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:562]:
```suggestion
3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This JSON parameters file is used by Azure Resource Manager (ARM) for runtime inputs.
```

File: docs/policy/authoring-guide.md [hunk 8]
```diff
@@ -0,0 +1,970 @@
...
+              "policyDefinitionManagementGroupId": {
+                  "value": "{{var-topLevelManagementGroupName}}"
+              }
+          }
+      }
+    ```
+
+#### **Step 2: Create policy set assignment template**
+
+1. Navigate to `policy/custom/assignments` and create two files.  Replace `POLICY_SET_ASSIGNMENT` with the name of your assignment such as `loganalytics`.
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:602]:
```suggestion
1. Navigate to `/policy/custom/assignments` and create two files.  Replace `POLICY_SET_ASSIGNMENT` with the name of your assignment such as `loganalytics`.
```

File: docs/policy/authoring-guide.md [hunk 9]
```diff
@@ -0,0 +1,970 @@
...
+        scope: managementGroup()
+        properties: {
+          roleDefinitionId: '/providers/Microsoft.Authorization/roleDefinitions/749f88d5-cbae-40b8-bcfc-e573ddc772fa'
+          principalId: policySetAssignment.identity.principalId
+          principalType: 'ServicePrincipal'
+        }
+      }
+    ```
+
+3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This parameters JSON file is used by Azure Resource Manager (ARM) for runtime inputs.
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:751]:
```suggestion
3. Edit the JSON parameters file to define the input parameters for the Bicep template.  This JSON parameters file is used by Azure Resource Manager (ARM) for runtime inputs.
```

File: docs/policy/authoring-guide.md [hunk 10]
```diff
@@ -0,0 +1,970 @@
...
+            "logAnalyticsResourceId": {
+                "value": "{{var-logging-logAnalyticsWorkspaceResourceId}}"
+            }
+          }
+      }
+    ```
+
+#### **Step 3: Configure Azure DevOps Pipeline**
+
+  * Edit `.pipelines/policy.yml`
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:803]:
```suggestion
  * Edit `/.pipelines/policy.yml`
```

File: docs/policy/authoring-guide.md [hunk 11]
```diff
@@ -0,0 +1,970 @@
...
+
+### Remove custom policy definition
+
+**Steps**
+
+* [Step 1: Remove policy definition](#step-1-remove-policy-definition)
+
+#### **Step 1: Remove policy definition**
+
+* Navigate to `policy/custom/definitions/policy` and identify the custom policy definition directory to delete.
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:898]:
```suggestion
* Navigate to `/policy/custom/definitions/policy` and identify the custom policy definition directory to delete.
```

File: docs/policy/authoring-guide.md [hunk 12]
```diff
@@ -0,0 +1,970 @@
...
+
+**Steps**
+  * [Step 1: Remove custom policy set definition](#step-1-remove-custom-policy-set-definition)
+  * [Step 2: Remove custom policy set assignment](#step-2-remove-custom-policy-set-assignment)
+  * [Step 3: Remove custom policy set from Azure DevOps Pipeline](#step-3-remove-custom-policy-set-from-azure-devops-pipeline)
+  * [Step 4: Remove custom policy set assignment's IAM assignments](#step-4-remove-custom-policy-set-assignments-iam-assignments)
+
+#### Step 1: Remove custom policy set definition
+
+* Navigate to `policy/custom/definitions/policyset` and delete the policy set definition Bicep template and JSON parameter files.
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:917]:
```suggestion
* Navigate to `/policy/custom/definitions/policyset` and delete the policy set definition Bicep template and JSON parameter files.
```

File: docs/policy/authoring-guide.md [hunk 13]
```diff
@@ -0,0 +1,970 @@
...
+
+#### Step 2: Remove custom policy set assignment
+
+* Navigate to `policy/custom/assignments` and delete the policy set assignment Bicep template and JSON parameter files.
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:921]:
```suggestion
* Navigate to `/policy/custom/assignments` and delete the policy set assignment Bicep template and JSON parameter files.
```

File: docs/policy/authoring-guide.md [hunk 14]
```diff
@@ -0,0 +1,970 @@
...
+
+
+#### Step 3: Remove custom policy set from Azure DevOps Pipeline
+
+  * Edit `.pipelines/policy.yml`
```
Comment (skeeler) [inline docs/policy/authoring-guide.md:926]:
```suggestion
  * Edit `/.pipelines/policy.yml`
```

**Review by skeeler — COMMENTED** [review]
Comments and suggestions inline. Also recommend installing & using the following 2 VS Code extensions to help lint markdown, detect & resolve warnings:
* https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint
* https://marketplace.visualstudio.com/items?itemName=docsmsft.docs-markdown

File: docs/policy/authoring-guide.md (hunk 1 above)
Comment (SenthuranSivananthan) [reply docs/policy/authoring-guide.md:174]: Done

**Review by skeeler — APPROVED** [review]

**Comment by hudua** [comment]
Policy authoring guide looks great, particularly the step-by-step instructions! 

One optional wording edit for the policy authoring guide, here's a suggestion for [Step 1 of collection info for built-in policy set assignment](https://github.com/Azure/CanadaPubSecALZ/blob/main/docs/policy/authoring-guide.md#step-1-collect-information) 

Suggest wording to put in front of Step 1, e.g. 

"We will not be assigning the policy through Azure Portal, but use these steps to identify the necessary info, such as name, definition ID, permissions, and parameters, which are required for the Policy Assignment" 

(instead of Step 1 - part 3) to provide motivation of using Azure Portal to collect info (and not to actually assign the policy set)

**Comment by SenthuranSivananthan** [comment]
Thanks for the feedback @hudua.  I'll send another PR with the improvements.

**Pull request merged**